sklearn/linear_model/_glm/distribution.py

```diff
@@ -0,0 +1,397 @@
+        """
+        return -2 * (y - mu) / self.unit_variance(mu)
+
+    def deviance(self, y, mu, weights=1):
+        r"""Compute the deviance.
+
+        The deviance is a weighted sum of the per sample unit deviances,
+        :math:`D = \sum_i s_i \cdot d(y_i, \mu_i)`
+        with weights :math:`s_i` and unit deviance :math:`d(y,\mu)`.
+        In terms of the log-likelihood it is :math:`D = -2\phi\cdot
+        \left(loglike(y,\mu,\frac{phi}{s})
+        - loglike(y,y,\frac{phi}{s})\right)`.
+
+        Parameters
+        ----------
+        y : array, shape (n_samples,)
+            Target values.
+
+        mu : array, shape (n_samples,)
+            Predicted mean.
+
+        weights : array, shape (n_samples,) (default=1)
+            Weights or exposure to which variance is inverse proportional.
+        """
+        return np.sum(weights * self.unit_deviance(y, mu))
+
+    def deviance_derivative(self, y, mu, weights=1):
+        """Compute the derivative of the deviance w.r.t. mu.
+
+        It gives :math:`\\frac{\\partial}{\\partial\\mu} D(y, \\mu; weights)`.
+
+        Parameters
+        ----------
+        y : array, shape (n_samples,)
+            Target values.
+
+        mu : array, shape (n_samples,)
+            Predicted mean.
+
+        weights : array, shape (n_samples,) (default=1)
+            Weights or exposure to which variance is inverse proportional.
+        """
+        return weights * self.unit_deviance_derivative(y, mu)
+
+    def _mu_deviance_derivative(self, coef, X, y, weights, link):
+        """Compute mu and the derivative of the deviance w.r.t coef."""
+        lin_pred = _safe_lin_pred(X, coef)
+        mu = link.inverse(lin_pred)
+        d1 = link.inverse_derivative(lin_pred)
+        temp = d1 * self.deviance_derivative(y, mu, weights)
+        if coef.size == X.shape[1] + 1:
+            devp = np.concatenate(([temp.sum()], temp @ X))
+        else:
+            devp = temp @ X  # sampe as X.T @ temp
+        return mu, devp
+
+
+class TweedieDistribution(ExponentialDispersionModel):
+    r"""A class for the Tweedie distribution.
+
+    A Tweedie distribution with mean :math:`\mu=\mathrm{E}[Y]` is uniquely
+    defined by it's mean-variance relationship
+    :math:`\mathrm{Var}[Y] \propto \mu^power`.
+
+    Special cases are:
+
+    ===== ================
+    Power Distribution
+    ===== ================
+    0     Normal
+    1     Poisson
+    (1,2) Compound Poisson
+    2     Gamma
+    3     Inverse Gaussian
+
+    Parameters
+    ----------
+    power : float (default=0)
+            The variance power of the `unit_variance`
+            :math:`v(\mu) = \mu^{power}`.
+            For ``0<power<1``, no distribution exists.
+    """
+    def __init__(self, power=0):
+        # validate power and set _upper_bound, _include_upper_bound attrs
+        self.power = power
+
+    @property
+    def power(self):
+        return self._power
+
+    @power.setter
+    def power(self, power):
```

Comment:
Yes that part is a bit tricky. For the code, what matters is `TweedieRegression.family.power` (where `family` is a TweedieDistribution that defines the deviance, derivatives etc). However with respect to the API the public parameter is `TweedieRegression.power`, we would want to be able to change this parameter e.g. in a grid search. Which means that if we want to subclass a single `GeneralizedLinearRegressor` class for all these special cases estimators, when we change est.power it also needs to change `est.family.power`.

On the other side, for instance `PoissonRegressor.family.power` also exists, since it's a particular case of the TweedieRegressor, however you don't want to allow to change the power there, because otherwise it means you could transform it to GammaRegressor or some other estimator.

Generally properties for `family` is a way to make this work while also satisfying the `check_estimator` or not setting extra attributes at __init__. I haven't found a better way of achieving this so far.  Alternatively it would mean that can't just subclass a single GLM class and need different implementations, which I find non ideal.

Reply:
Ok that makes sense. Would this be a non-issue if we only exposed `GeneralizedLinearRegressor`? (I'm not sure I completely understand the relation between the families, `power`, and the estimator classes yet).

Reply:
@rth I think your explanation is good and should be reflected in a comment in the code.